pkg/disttask/framework/metering/metering.go

```diff
@@ -0,0 +1,262 @@
+func RecordDXFWriteDataBytes(store kv.Storage, taskType TaskType, taskID int64, size uint64) {
+	if !kv.IsUserKS(store) || meteringInstance.Load() == nil {
+		return
+	}
+	userKS := store.GetKeyspace()
+	meteringInstance.Load().Record(userKS, &Data{
+		taskType:       string(taskType),
+		taskID:         taskID,
+		writeDataBytes: size,
+	})
+}
+
+// GetMetering gets the metering instance.
+func GetMetering() (*Meter, error) {
+	v := meteringInstance.Load()
+	if v == nil {
+		return nil, errors.New("metering instance is not initialized")
+	}
+	return v, nil
+}
+
+// SetMetering sets the metering instance for dxf.
+func SetMetering(m *Meter) {
+	meteringInstance.Store(m)
+}
+
+// Data represents the metering data.
+type Data struct {
+	putRequests    uint64
+	getRequests    uint64
+	readDataBytes  uint64
+	writeDataBytes uint64
+
+	taskType string
+	taskID   int64
+}
+
+func (m *Data) merge(other *Data) {
+	m.putRequests += other.putRequests
+	m.getRequests += other.getRequests
+	m.readDataBytes += other.readDataBytes
+	m.writeDataBytes += other.writeDataBytes
+}
+
+// Meter is responsible for recording and reporting metering data.
+type Meter struct {
+	sync.Mutex
+	ctx    context.Context
+	data   map[string]Data // keyspace -> meter data
+	uuid   string
+	writer *meteringwriter.MeteringWriter
+	logger *zap.Logger
+}
+
+// NewMeter creates a new Meter instance.
+func NewMeter(cfg *mconfig.MeteringConfig) (*Meter, error) {
+	logger := logutil.BgLogger().With(zap.String("component", "meter"))
+	if len(cfg.Bucket) == 0 {
+		return nil, nil
+	}
+	if cfg.AWS == nil {
+		return nil, nil
+	}
+	providerType := storage.ProviderTypeS3
+	if len(cfg.Type) > 0 {
+		providerType = cfg.Type
+	}
+
+	s3Config := &storage.ProviderConfig{
+		Type:   providerType,
+		Bucket: cfg.Bucket,
+		Region: cfg.Region,
+		Prefix: cfg.Prefix,
+		AWS: &storage.AWSConfig{
+			AssumeRoleARN: cfg.AWS.AssumeRoleARN,
+		},
+	}
+	provider, err := storage.NewObjectStorageProvider(s3Config)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create storage provider")
+	}
+	meteringConfig := mconfig.DefaultConfig().WithLogger(logger)
+	writer := meteringwriter.NewMeteringWriter(provider, meteringConfig)
+	return &Meter{
+		logger: logger,
+		data:   make(map[string]Data),
+		writer: writer,
+		uuid:   strings.ReplaceAll(uuid.New().String(), "-", "_"), // no dash in the S3 path
+	}, nil
+}
+
+// Record records a meter data.
+func (m *Meter) Record(userKeyspace string, other *Data) {
+	m.Lock()
+	defer m.Unlock()
+	orig := m.data[userKeyspace]
+	orig.merge(other)
+	m.data[userKeyspace] = orig
+}
+
+// StartFlushLoop creates a flush loop.
+func (m *Meter) StartFlushLoop(ctx context.Context) {
+	// Control the writing timestamp accurately enough so that the previous round won't be overwritten by the next round.
+	curTime := time.Now().Unix()
+	nextTime := curTime/writeInterval*writeInterval + writeInterval
+	for ctx.Err() == nil {
```

Comment:
```suggestion
	for {
```
maybe just break on ctx done